File: LinkedList-Palindrome/primeira.py
from re import L


class ListNode:
    def __init__(self,val:int = 0, _next:'ListNode' = None):
        self.val = val
        self.next = _next
    def __repr__(self) -> None:
        cur = self
        value = ""
        while cur != None:
            value += f"{cur.val} "
            cur = cur.next
        return value


#O(n²) -> time ; O(n) -> space
def isPalindromeNaive(head:'ListNode') -> bool:
    size = 0
    pointer = head
    while pointer != None:
        size+= 1
        pointer = pointer.next

    metade = []
    cur = head
    for i in range(size//2):
        metade.insert(0,cur.val)
        cur = cur.next
    if(size%2 != 0): cur = cur.next
    for i in metade:
        if cur.val != i:
            return False
        cur = cur.next
    return True

#O(n) -> time ; O(1) -> space
def isPalindrome(head:'ListNode') -> bool:
    slowP = head
    fastP = head
    while fastP != None and fastP.next != None:
        fastP = fastP.next.next
        slowP = slowP.next

    if fastP != None : slowP = slowP.next

    #inverter head
    prev: 'ListNode' = None
    _next: 'ListNode'

    while slowP != None:
        _next = slowP.next
        slowP.next = prev
        prev = slowP
        slowP = _next
    slowP = prev

    while slowP != None:
        if slowP.val != head.val:
            return False
        slowP = slowP.next
        head = head.next
    return True

File: LinkedList-Palindrome/primeira.cpp
#include<iostream>
#include<list>
#include<algorithm>

class ListNode{
    public:
        int val;
        ListNode* next;
        ListNode():val(0),next(nullptr){}
        ListNode(int val):val(val), next(nullptr){}
        ListNode(int val, ListNode* next):val(val), next(next){}
};

void printList(ListNode* head){
    ListNode* cur = head;
    while(cur != nullptr) {
        std::cout << cur->val << " ";
        cur = cur->next;
    }
    std::cout << "\n";
}

void printList(std::list<int> valores){
    for(int var : valores){
        std::cout << var << " ";
    }
    std::cout << "\n";
}

bool isPalindromeNaive(ListNode* head){
    ListNode* pointer = head;
    int size = 0;
    std::list<int> metade;
    while(pointer != nullptr){
        pointer= pointer->next;
        size++;
    }
    pointer = head;
    for(int i = 0; i < size/2; i++){
        metade.insert(metade.begin(), pointer->val);
        pointer = pointer->next;
    }
    if(size%2 != 0) pointer = pointer->next;
    for(int valor : metade){
        if(valor != pointer->val) return false;
        pointer = pointer->next;
    }
    return true;
}

bool isPalindrome(ListNode* head){
    ListNode* slowP = head;
    ListNode* fastP = head;
    while(fastP != nullptr && fastP->next != nullptr){
        slowP = slowP->next;
        fastP = fastP->next->next;
    }

    if(fastP!=nullptr) slowP = slowP->next;

    ListNode* prev = nullptr;
    ListNode* next;
    while(slowP != nullptr){
        next = slowP->next;
        slowP->next = prev;
        prev = slowP;
        slowP = next;
    }
    slowP = prev;
    while(slowP != nullptr){
        if(slowP->val != head->val) return false;
        slowP = slowP->next;
        head = head->next;
    }

    return true;
}

int main(){
    std::cout << isPalindrome(new ListNode(1, new ListNode(2, new ListNode(2, new ListNode(1))))) << "\n";
    std::cout << isPalindrome(new ListNode(1, new ListNode(3, new ListNode(2, new ListNode(1))))) << "\n";
    std::cout << isPalindrome(new ListNode(1, new ListNode(2, new ListNode(1)))) << "\n";
    return 0;
}
